target/whirl.c

```diff
@@ -0,0 +1,1221 @@
+}
+
+static RingDirection reverse_direction(RingDirection dir) {
+    return dir == CLOCKWISE ? COUNTERCLOCKWISE : CLOCKWISE;
+}
+
+// Adds a 1 instruction to the current code segment, and update the current
+// ring state to account for it
+static void emit_one(WhirlCodeSegment *segment, RingState *state) {
+    if (state->active_ring == MATH_RING) {
+        if (state->math_dir == CLOCKWISE) {
+            state->cur_math_pos++;
+            if (state->cur_math_pos == NUM_CMDS_PER_RING) {
+                state->cur_math_pos = 0;
+            }
+        }
+        else {
+            if (state->cur_math_pos == 0) {
+                state->cur_math_pos = NUM_CMDS_PER_RING;
+            }
+            state->cur_math_pos--;
+        }
+    }
+    else {
+        if (state->op_dir == CLOCKWISE) {
+            state->cur_op_pos++;
+            if (state->cur_op_pos == NUM_CMDS_PER_RING) {
+                state->cur_op_pos = 0;
+            }
+        }
+        else {
+            if (state->cur_op_pos == 0) {
+                state->cur_op_pos = NUM_CMDS_PER_RING;
+            }
+            state->cur_op_pos--;
+        }
+    }
+
+    state->last_is_zero = false;
+    emit_instruction(segment, '1');    
+}
+
+// Adds a 0 instruction to the current code segment, and update the current
+// ring state to account for it
+static void emit_zero(WhirlCodeSegment *segment, RingState *state) {
+    if (state->active_ring == MATH_RING) {
+        state->math_dir = reverse_direction(state->math_dir);
+    }
+    else {
+        state->op_dir = reverse_direction(state->op_dir);
+    }
+
+    if (state->last_is_zero) {
+        state->active_ring = (state->active_ring == MATH_RING ? OPERATION_RING :
+                                                                MATH_RING);
+    }
+
+    state->last_is_zero = !state->last_is_zero;
+    emit_instruction(segment, '0');
+}
+
+// Returns whether we should switch the direction of the current ring, based on
+// which direction would generate shorter code for the target command we want
+static bool should_switch(int cur_pos, int target_pos, RingDirection dir) {
+    int ones_required;
+
+    if (dir == CLOCKWISE) {
+        ones_required = target_pos - cur_pos;
+    }
+    else {
+        ones_required = cur_pos - target_pos;
+    }
+
+    if (ones_required < 0) {
+        ones_required += NUM_CMDS_PER_RING;
+    }
+
+    return ones_required > (NUM_CMDS_PER_RING / 2);
+}
+
+static void generate_math_command(WhirlCodeSegment *segment, RingState *state, MathCmd cmd);
+
+// Generates the Whirl code to execute a command on the operations ring, taking
+// into account the current state of the command rings
+static void generate_op_command(WhirlCodeSegment *segment, RingState *state, OpCmd cmd) {
+    if (state->active_ring == MATH_RING) {
+        // If we're on the wrong ring, we need to execute a nop command to switch to
+        // the operations ring
+        generate_math_command(segment, state, OP_NOOP);
```

Comment:
MATH_NOOP?

https://travis-ci.org/shinh/elvm/builds/645138426#L329

Reply:
Whoops. I have fixed this now, along with the other instance where OP_NOOP and MATH_NOOP were switched.